code/game/jobs/job_exp.dm

```diff
@@ -0,0 +1,207 @@
+			else
+				var/xp_req = job.exp_requirements
+				if(config.use_exp_restrictions_heads_hours && ((job.title in command_positions) || job.title == "AI"))
+					xp_req = config.use_exp_restrictions_heads_hours * 60
+				jobs_locked += job.title + " (" + get_exp_format(text2num(play_records[job.exp_type])) + " / " + get_exp_format(xp_req) + " [job.exp_type] EXP)"
+	if(jobs_unlocked.len)
+		return_text += "<BR><BR>Jobs Unlocked:<UL> "
+		for(var/text in jobs_unlocked)
+			return_text += "<LI>[text]</LI>"
+		return_text += "</UL>"
+	if(jobs_locked.len)
+		return_text += "<BR><BR>Jobs Not Unlocked:<UL>"
+		for(var/text in jobs_locked)
+			return_text += "<LI>[text]</LI>"
+		return_text += "</UL>"
+	return return_text
+
+
+/client/proc/get_exp_general()
+	var/list/play_records = params2list(prefs.exp)
+	var/exp_gen = text2num(play_records[EXP_TYPE_CREW])
+	return get_exp_format(exp_gen)
+
+/proc/get_exp_format(var/expnum)
+	if(expnum > 60)
+		return num2text(round(expnum / 60)) + "h"
+	else if(expnum > 0)
+		return num2text(expnum) + "m"
+	else
+		return "0h"
+
+/proc/update_exp(var/minutes, var/announce_changes = 0)
+	if(!establish_db_connection())
+		return -1
+	spawn(0)
+		for(var/client/C in clients)
+			if(C.inactivity >= (10 MINUTES))
+				continue
+
+			var/DBQuery/exp_read = dbcon.NewQuery("SELECT exp FROM [format_table_name("player")] WHERE ckey='[C.ckey]'")
+			exp_read.Execute()
+			var/list/read_records = list()
+			while(exp_read.NextRow())
+				read_records = params2list(exp_read.item[1])
+
+			var/list/play_records = list()
+			for(var/rtype in exp_jobsmap)
+				if(text2num(read_records[rtype]))
+					play_records[rtype] = text2num(read_records[rtype])
+				else
+					play_records[rtype] = 0
+			if(C.mob.stat == CONSCIOUS && C.mob.mind.assigned_role)
+				play_records[EXP_TYPE_CREW] += minutes
+				if(announce_changes)
+					to_chat(C.mob,"<span class='notice'>You got: [minutes] General EXP!")
+				for(var/cat in exp_jobsmap)
+					if(exp_jobsmap[cat]["titles"])
+						if(C.mob.mind.assigned_role in exp_jobsmap[cat]["titles"])
+							play_records[cat] += minutes
+							if(announce_changes)
+								to_chat(C.mob,"<span class='notice'>You got: [minutes] [cat] EXP!")
+				if(C.mob.mind.special_role)
+					play_records[EXP_TYPE_SPECIAL] += minutes
+					if(announce_changes)
+						to_chat(C.mob,"<span class='notice'>You got: [minutes] Special EXP!")
+			else if(isobserver(C.mob))
+				play_records[EXP_TYPE_GHOST] += minutes
+				if(announce_changes)
+					to_chat(C.mob,"<span class='notice'>You got: [minutes] Ghost EXP!")
+			else
+				return
+			var/new_exp = list2params(play_records)
+			C.prefs.exp = new_exp
+			new_exp = sanitizeSQL(new_exp)
+			var/DBQuery/update_query = dbcon.NewQuery("UPDATE [format_table_name("player")] SET exp = '[new_exp]' WHERE ckey='[C.ckey]'")
+			update_query.Execute()
+			sleep(10)
+
+
+/hook/roundstart/proc/exptimer()
+	if(!config.sql_enabled || !config.use_exp_tracking)
+		return  0
```

Comment:
Hook procs like this should `return 1`, or the hook handler will report an error. You should also `return 1` at the end of this proc.